File: lbr_demos/lbr_demos_py/lbr_demos_py/wrench_sine_overlay.py
import math
from copy import deepcopy

import rclpy
from rcl_interfaces.srv import GetParameters
from rclpy.node import Node

# import lbr_fri_idl
from lbr_fri_idl.msg import LBRState, LBRWrenchCommand


class WrenchSineOverlayNode(Node):
    def __init__(self, node_name: str) -> None:
        super().__init__(node_name)

        self._amplitude_x, self._amplitude_y = 5.0, 5.0  # N
        self._frequency_x, self._frequency_y = 0.25, 0.25  # Hz
        self._phase_x, self._phase_y = 0.0, 0.0
        self._lbr_wrench_command = LBRWrenchCommand()

        # create publisher to command/wrench
        self._lbr_wrench_command_pub = self.create_publisher(
            LBRWrenchCommand, "command/wrench", 1
        )

        # create subscription to state
        self._lbr_state = None
        self._lbr_state_sub = self.create_subscription(
            LBRState, "state", self._on_lbr_state, 1
        )

        # get control rate from controller_manager
        self._dt = None
        self._retrieve_update_rate()

    def _on_lbr_state(self, lbr_state: LBRState) -> None:
        if self._dt is None:
            return
        if self._lbr_state is None:
            self._lbr_state = lbr_state
        self._lbr_wrench_command.joint_position = deepcopy(
            self._lbr_state.measured_joint_position
        )

        if lbr_state.session_state == 4:  # KUKA::FRI::COMMANDING_ACTIVE == 4
            # overlay wrench sine wave on x / y direction
            self._lbr_wrench_command.wrench[0] = self._amplitude_x * math.sin(
                self._phase_x
            )
            self._lbr_wrench_command.wrench[1] = self._amplitude_y * math.sin(
                self._phase_y
            )
            self._phase_x += 2 * math.pi * self._frequency_x * self._dt
            self._phase_y += 2 * math.pi * self._frequency_y * self._dt

            self._lbr_wrench_command_pub.publish(self._lbr_wrench_command)
        else:
            # reset phase
            self._phase_x, self._phase_y = 0.0, 0.0

    def _retrieve_update_rate(self) -> float:
        paramter_client = self.create_client(
            GetParameters, "controller_manager/get_parameters"
        )
        paramter_name = "update_rate"
        while not paramter_client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                raise RuntimeError("Interrupted while waiting for service.")
            self.get_logger().info(f"Waiting for {paramter_client.srv_name}...")
        future = paramter_client.call_async(
            GetParameters.Request(names=[paramter_name])
        )
        rclpy.spin_until_future_complete(self, future)
        if future.result() is None:
            raise RuntimeError(f"Failed to get parameter '{paramter_name}'.")
        update_rate = future.result().values[0].integer_value
        self.get_logger().info(f"{paramter_name}: {update_rate} Hz")
        self._dt = 1.0 / float(update_rate)


def main(args: list = None) -> None:
    rclpy.init(args=args)
    rclpy.spin(WrenchSineOverlayNode("wrench_sine_overlay_node"))
    rclpy.shutdown()

File: lbr_demos/lbr_demos_cpp/src/wrench_sine_overlay.cpp
#include <functional>
#include <math.h>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"

// include fri for session state
#include "friClientIf.h"

// include lbr_fri_idl
#include "lbr_fri_idl/msg/lbr_state.hpp"
#include "lbr_fri_idl/msg/lbr_wrench_command.hpp"
#include "lbr_fri_ros2/utils.hpp"

class WrenchSineOverlay {
  static constexpr double amplitude_x_ = 5.;   // N
  static constexpr double amplitude_y_ = 5.;   // N
  static constexpr double frequency_x_ = 0.25; // Hz
  static constexpr double frequency_y_ = 0.25; // Hz

public:
  WrenchSineOverlay(const rclcpp::Node::SharedPtr node) : node_(node), phase_x_(0.), phase_y_(0.) {
    // create publisher to command/wrench
    lbr_wrench_command_pub_ =
        node_->create_publisher<lbr_fri_idl::msg::LBRWrenchCommand>("command/wrench", 1);

    // create subscription to state
    lbr_state_sub_ = node_->create_subscription<lbr_fri_idl::msg::LBRState>(
        "state", 1, std::bind(&WrenchSineOverlay::on_lbr_state_, this, std::placeholders::_1));

    // get control rate from controller_manager
    auto update_rate =
        lbr_fri_ros2::retrieve_paramter(node_, "controller_manager", "update_rate").as_int();
    dt_ = 1.0 / static_cast<double>(update_rate);
  };

protected:
  void on_lbr_state_(const lbr_fri_idl::msg::LBRState::SharedPtr lbr_state) {
    if (!lbr_state_init_) {
      lbr_state_ = *lbr_state;
      lbr_state_init_ = true;
    }
    lbr_wrench_command_.joint_position = lbr_state_.measured_joint_position;

    if (lbr_state->session_state == KUKA::FRI::COMMANDING_ACTIVE) {
      // overlay wrench sine wave on x / y direction
      lbr_wrench_command_.wrench[0] = amplitude_x_ * sin(phase_x_);
      lbr_wrench_command_.wrench[1] = amplitude_y_ * sin(phase_y_);
      phase_x_ += 2 * M_PI * frequency_x_ * dt_;
      phase_y_ += 2 * M_PI * frequency_y_ * dt_;

      lbr_wrench_command_pub_->publish(lbr_wrench_command_);
    } else {
      // reset phase
      phase_x_ = 0.;
      phase_y_ = 0.;
    }
  };

protected:
  rclcpp::Node::SharedPtr node_;
  double dt_;
  double phase_x_, phase_y_;
  rclcpp::Publisher<lbr_fri_idl::msg::LBRWrenchCommand>::SharedPtr lbr_wrench_command_pub_;
  rclcpp::Subscription<lbr_fri_idl::msg::LBRState>::SharedPtr lbr_state_sub_;
  bool lbr_state_init_ = false;
  lbr_fri_idl::msg::LBRState lbr_state_;
  lbr_fri_idl::msg::LBRWrenchCommand lbr_wrench_command_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<rclcpp::Node>("wrench_sine_overlay");
  WrenchSineOverlay wrench_sine_overlay(node);
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
};
